test/penalizer/Penalizer.test.ts

```diff
@@ -0,0 +1,366 @@
+      const receipt = relayHelper.createReceipt({ relayRequest: rr, signature: sig })
+
+      await expectRevert(
+        penalizer.claim(receipt),
+        'worker signature mismatch'
+      )
+    })
+
+    it('due to forged signature', async function () {
+      const rr = await relayHelper.createRelayRequest({
+        from: sender.address,
+        to: recipient.address,
+        relayData: '0xdeadbeef06',
+        enableQos: true
+      })
+      const sig = relayHelper.getRelayRequestSignature(rr, sender)
+
+      const receipt = relayHelper.createReceipt({ relayRequest: rr, signature: sig })
+      receipt.workerSignature = web3.utils.randomHex(130)
+
+      await expectRevert(
+        penalizer.claim(receipt),
+        'worker signature mismatch'
+      )
+    })
+
+    it('due to worker address mismatch', async function () {
+      const rr = await relayHelper.createRelayRequest({
+        from: sender.address,
+        to: recipient.address,
+        relayData: '0xdeadbeef07',
+        enableQos: true
+      })
+      const sig = relayHelper.getRelayRequestSignature(rr, sender)
+
+      const receipt = relayHelper.createReceipt({ relayRequest: rr, signature: sig })
+      receipt.commitment.relayWorker = otherAccount
+      await relayHelper.signReceipt(receipt)
+
+      await expectRevert(
+        penalizer.claim(receipt),
+        'worker address does not match'
+      )
+    })
+
+    it('due to relay hub mismatch', async function () {
+      const rr = await relayHelper.createRelayRequest({
+        from: sender.address,
+        to: recipient.address,
+        relayData: '0xdeadbeef07',
+        enableQos: true
+      })
+      const sig = relayHelper.getRelayRequestSignature(rr, sender)
+
+      const receipt = relayHelper.createReceipt({ relayRequest: rr, signature: sig })
+      receipt.commitment.relayHubAddress = otherAccount
+      await relayHelper.signReceipt(receipt)
+
+      await expectRevert(
+        penalizer.claim(receipt),
+        'relay hub does not match'
+      )
+    })
+
+    it('due to claim not being made by commitment receiver', async function () {
+      const rr = await relayHelper.createRelayRequest({
+        from: sender.address,
+        to: recipient.address,
+        relayData: '0xdeadbeef08',
+        enableQos: true
+      })
+      const sig = relayHelper.getRelayRequestSignature(rr, sender)
+
+      const receipt = relayHelper.createReceipt({ relayRequest: rr, signature: sig })
+      await relayHelper.signReceipt(receipt)
+
+      await expectRevert(
+        penalizer.claim(receipt, { from: otherAccount }),
+        'receiver must claim commitment'
+      )
+    })
+  })
+
+  describe('should receive claims', function () {
+    before(async function () {
+      // sender will need to be funded to complete claim calls
+      await fundAccount(fundedAccount, sender.address, '10')
+    })
+
+    // from this point onwards txs must be put together manually since the sender account is locked
```

Comment:
a little bit more context to this:

for starters, it's more convenient for the account which needs relaying to have no balance, since this is the whole point of the relaying system. so using one of the 10 unlocked truffle accounts isn't a great idea to start with.

but more importantly, we need an account with an accessible private key in order to create a smart wallet with it as its owner (see line `53`).

the only way to do this with truffle is to use a seed phrase and hard-code the (predicted) private key for one of the accounts, but i'm not thrilled at this idea in terms of coding practices.

as a result, we'll be using a manually generated gasless account. this will require `Transaction` objects to be created and signed manually, since direct contract calls won't work when setting the `from` field using the `sender` account (as it is locked as far as truffle is concerned).